gui/rendering/Node.cpp

```diff
@@ -0,0 +1,160 @@
+//
+// Created by hcaramizaru on 4/15/24.
+//
+
+#include "Node.hpp"
+#include "sme/logger.hpp"
+
+namespace rendering {
+
+Node::Node(const std::string &name, const QVector3D &position,
+           const QVector3D &rotation, const QVector3D &scale)
+    : name(name), m_position(position), m_rotation(rotation), m_scale(scale) {
+
+  this->worldTransform.setToIdentity();
+  this->localTransform.setToIdentity();
+}
+
+Node::Node(const std::string &name, const QVector3D &position,
+           const QVector3D &rotation)
+    : Node(name, position, rotation, QVector3D(1, 1, 1)) {}
+
+Node::Node(const std::string &name, const QVector3D &position)
+    : Node(name, position, QVector3D(0, 0, 0)) {}
+
+Node::Node(const std::string &name) : Node(name, QVector3D(0, 0, 0)) {}
+
+Node::Node() : Node("Node") {}
+
+Node::~Node() { SPDLOG_DEBUG("Removing Node \"" + name + "\""); }
+
+void Node::update(float delta) {
+  // It should be overwritten in case extra computation is required per Node.
+}
+
+void Node::updateWorldTransform(float delta) {
+  if (m_dirty) {
+    auto parent_ref = this->parent.lock();
+    this->updateLocalTransform();
+    if (parent_ref.get() != nullptr) {
+      worldTransform = parent_ref.get()->worldTransform * localTransform;
+    } else {
+      worldTransform = localTransform;
+    }
+
+    m_dirty = false;
+    for (const auto &node : children) {
+      node.get()->markDirty();
+      node.get()->updateWorldTransform(delta);
+    }
+  } else {
+    for (const auto &node : children) {
+      node.get()->updateWorldTransform(delta);
+    }
+  }
+  this->update(delta);
+}
+
+// TODO: Use quaternions as a substitute to euler angles
+void Node::updateLocalTransform() {
+  if (m_dirty) {
+
+    localTransform.setToIdentity();
+
+    localTransform.translate(m_position);
+
+    localTransform.scale(m_scale);
+
+    // check which Euler angle convention it is used
+    localTransform.rotate(m_rotation.z(), 0.0f, 0.0f, 1.0f);
+    localTransform.rotate(m_rotation.y(), 0.0f, 1.0f, 0.0f);
+    localTransform.rotate(m_rotation.x(), 1.0f, 0.0f, 0.0f);
+  }
+}
+
+void Node::add(std::shared_ptr<Node> node, bool transformInLocalSpace) {
+
+  // TODO: Implement use case with node transform in global coordinates.
+  assert(transformInLocalSpace == true);
+
+  if (node) {
+    node.get()->parent = shared_from_this();
+    node.get()->m_dirty = true;
+    children.push_back(node);
+  }
+}
+
+void Node::remove() {
+
+  auto parent_ref = this->parent.lock();
+  if (parent_ref.get() == nullptr) {
+    SPDLOG_WARN("Attempted to remove \"" + name + "\", but it has no parent");
+    return;
+  }
+
+  auto it = std::find_if(parent_ref.get()->children.begin(),
+                         parent_ref.get()->children.end(),
+                         [this](auto i) { return i.get() == this; });
+  if (it != parent_ref.get()->children.end()) {
+    parent_ref.get()->children.erase(it);
+  }
```

Comment:
this can be simpler using erase_if:
`std::erase_if(parent_ref->children, [this](auto child) { return child.get() == this; });`